### DISCUSSION TITLE
Webpack can't dynamically import a file at runtime

### DISCUSSION THREAD

**Discussion by mwaqar666** [nrwl/nx#13862] [General]
I have created an Nx monorepo containing around 9 libraries *(all using `@nrwl/js:tsc` for building)* and an application *(using `@nrwl/webpack:webpack`)*.

The application exposes a command line library which is used to run database migration files all over the Nx monorepo. The identity of database migration files is that, after building, they ends with `.migration.js`. I am using glob pattern to find these files all over in libs folder and then I dynamically import these files. 

Now, here's the buggy part: These migration files can be anywhere in the libs folder. Suppose if the command line library dependency path is like this: `Cmd Lib -> Db -> Config -> Utils`, then only these four libs gets build on building the `Cmd Lib`, but the migration file can be in the other project, say `DAL`, and obviously `DAL` was not built, so I have to build it manually. But even after building it manually, the dynamic import can't resolve the built files `*.migraqtion.js` in the `DAL` library, stating the error that `Module can't be found` even though the file is there in the filesystem, I have manually build the project.

I have asked this on the webpack discussion [here](https://github.com/webpack/webpack/discussions/16566), but the guy asked me to ask on the Nx discussion instead.
The reproduction steps are:

1. Clone the repo: [https://github.com/mwaqar666/sca](https://github.com/mwaqar666/sca)
2. Switch to branch `backend/backend-integration`
3. Run `npx nx run sca-cmd:build` (Build the command-line application using Webpack)
4. Run `npx nx run backend-data-access-layer:build` (Build the project using `tsc` compiler from which files are imported dynamically)
5. Run `npx nx run sca-cmd:run:dev --verbose -- migration:up` (This simply executes the sca-cmd app using node)

I have added an `.env.dev` file to run the application. It contains nothing confidential or secret for now. I have also added the database drivers, in case someone tries to reproduce the issue. Just configure the values in `config/backend/.env.dev` and the database will be connected.

**Comment by bobalazek** [comment]
Hey @mwaqar666, did you ever found a solution for this? I'm currently having the same issue.

**Reply by mwaqar666** [reply]
Nope unfortunately :( I had other high priority tasks in the queue, so I didn't solved it yet. But I'll get back to it soon.

Let me know if you've think of any potential solution. Maybe it also point me in some right direction.

**Reply by bobalazek** [reply]
The closest I've come is just readdir() the dist directory of that library for *.js (or *.cjs) files and then doing a relative import() inside the CLI app. It's more of a hacky workaround, but it does the job for now, although I'm not sure if it's really suitable for all cases, especially if that included file links to other local imports - haven't tried that now.

**Comment by abhinav580** [comment]
I am also facing the same issue. Did anyone found any solution
